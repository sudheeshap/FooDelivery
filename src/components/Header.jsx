import React, { useState } from 'react';
import { Link } from 'react-router-dom';

export default function Header() {
  const [isLoggedIn] = useState(false);

  return (
    <header>
      <nav className="navbar">
        <div className="navbar__section">
          <div className="navbar__item">
            <Link to="/">
              <img
                className="navbar__logo"
                src="https://tailwindui.com/img/logos/workflow-mark-indigo-500.svg"
                alt="Workflow"
              />
            </Link>
          </div>
        </div>
        <div className="navbar__section">
          <div className="navbar__item">
            <div className={`navbar__cart ${isLoggedIn ? '' : 'navbar__cart--disabled'}`}>
              <i className="bi-bag" />
            </div>
          </div>

          {isLoggedIn ? (
            <div className="navbar__item">
              <div
                className="navbar__user"
                style={{
                  backgroundImage:
                    'url(https://images.unsplash.com/photo-1472099645785-5658abf4ff4e?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=facearea&facepad=2&w=256&h=256&q=80',
                }}
              >
                <img alt="" />
              </div>
            </div>
          ) : (
            <div className="navbar__item">
              <span className="navbar__link">Login</span>
            </div>
          )}
        </div>
      </nav>
    </header>
  );
}
