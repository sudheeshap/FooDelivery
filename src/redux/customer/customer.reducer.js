import { createSlice } from '@reduxjs/toolkit';

import INITIAL_STATE from '../initial-state';

const customerSlice = createSlice({
  name: 'customer',
  initialState: INITIAL_STATE.customer,
  reducers: {
    updateCustomer(state, action) {
      return action.payload.customer;
    },
  },
});

export const { updateCustomer } = customerSlice.actions;

export default customerSlice.reducer;
