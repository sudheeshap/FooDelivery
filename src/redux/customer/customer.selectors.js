import { createSelector } from '@reduxjs/toolkit';

/**
 * Select customer state
 */
const selectCustomerState = (state) => state.menuGroup;

/**
 * Select customer ID
 */
export const selectCustomerId = createSelector(
  [selectCustomerState],
  (customer) => customer.id,
);

/**
 * Select state: customer logged in?
 */
export const selectCustomerIsLoggedIn = createSelector(
  [selectCustomerId],
  (id) => !!id,
);

/**
 * Select customer mobile
 */
export const selectCustomerMobile = createSelector(
  [selectCustomerState],
  (customer) => customer.mobile,
);

/**
 * Select customer email
 */
export const selectCustomerEmail = createSelector(
  [selectCustomerState],
  (customer) => customer.email,
);
