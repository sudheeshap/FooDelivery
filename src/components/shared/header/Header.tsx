import React, { MouseEvent } from 'react';
import { useSelector } from 'react-redux';
import { Link, useHistory } from 'react-router-dom';

import {
  selectCartItemCount,
  selectCartRestaurant,
} from '../../../redux/cart/cart.selectors';
import { selectCustomerIsLoggedIn } from '../../../redux/customer/customer.selectors';
import { useAuth } from '../../../hooks/useAuth';
import CartIcon from '../../CartIcon';

export default function Header() {
  const { logout } = useAuth();
  const history = useHistory();

  const cartItemCount = useSelector(selectCartItemCount);
  const cartRestaurant = useSelector(selectCartRestaurant);
  const isLoggedIn = useSelector(selectCustomerIsLoggedIn);

  /**
   * Clicked on logout button
   */
  const handleClickLogout = (event: MouseEvent<HTMLAnchorElement>) => {
    event.preventDefault();

    // Logout customer
    logout();
  };

  /**
   * Clicked on cart icon
   */
  const handleClickCart = () => {
    history.push(`/restaurant/${cartRestaurant.slug}`);
  };

  return (
    <header className="header">
      <nav className="navbar">
        <div className="navbar__section">
          <div className="navbar__item">
            <Link to="/">
              <img
                className="navbar__logo"
                src="https://tailwindui.com/img/logos/workflow-mark-indigo-500.svg"
                alt="Workflow"
              />
            </Link>
          </div>
        </div>
        <div className="navbar__section">
          {isLoggedIn ? (
            <div className="navbar__item">
              {/* <div
                className="navbar__user"
                style={{
                  backgroundImage:
                    'url(https://images.unsplash.com/photo-1472099645785-5658abf4ff4e?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=facearea&facepad=2&w=256&h=256&q=80',
                }}
              >
                <img alt="" />
              </div> */}
              <a
                href="/"
                className="navbar__link"
                data-testid="link-logout"
                onClick={handleClickLogout}
              >
                Logout
              </a>
            </div>
          ) : (
            <div className="navbar__item">
              <Link
                to="/login"
                className="navbar__link"
                data-testid="link-login"
              >
                Login
              </Link>
            </div>
          )}

          <div className="navbar__item">
            <CartIcon
              count={cartItemCount}
              onClick={handleClickCart}
              data-testid="link-cart"
            />
          </div>
        </div>
      </nav>
    </header>
  );
}
