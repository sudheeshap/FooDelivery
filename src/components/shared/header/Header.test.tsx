import React from 'react';
import { mount, ReactWrapper } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { mountToJson } from 'enzyme-to-json';
import { BrowserRouter } from 'react-router-dom';

import Header from './Header';
import INITIAL_STATE from '../../../redux/initial-state';
import * as selectors from '../../../redux/customer/customer.selectors';

let headerWrapper: ReactWrapper;
const mockLogout = jest.fn();

jest.mock('../../../hooks/useAuth.js', () => ({
  useAuth: () => ({
    logout: mockLogout,
  }),
}));

const wrapperFactory = () => {
  const mockStore = configureStore([]);
  const store = mockStore(INITIAL_STATE);

  return mount(
    <BrowserRouter>
      <Provider store={store}>
        <Header />
      </Provider>
    </BrowserRouter>,
  );
};

describe('Header', () => {
  afterEach(() => {
    if (headerWrapper) {
      headerWrapper.unmount();
    }
  });

  it('renders the header', () => {
    headerWrapper = wrapperFactory();

    expect(mountToJson(headerWrapper)).toMatchSnapshot();
  });

  it('displays login link if user not logged in', () => {
    jest.spyOn(selectors, 'selectCustomerIsLoggedIn').mockReturnValue(false);

    headerWrapper = wrapperFactory();

    expect(
      headerWrapper.find('[data-testid="link-login"]').exists(),
    ).toBeTruthy();

    expect(
      headerWrapper.find('[data-testid="link-logout"]').exists(),
    ).toBeFalsy();
  });

  it('displays logout link when user is logged in', () => {
    jest.spyOn(selectors, 'selectCustomerIsLoggedIn').mockReturnValue(true);

    headerWrapper = wrapperFactory();

    expect(
      headerWrapper.find('[data-testid="link-logout"]').exists(),
    ).toBeTruthy();

    expect(
      headerWrapper.find('[data-testid="link-login"]').exists(),
    ).toBeFalsy();
  });

  it('should call logout() when clicking logout link', () => {
    jest.spyOn(selectors, 'selectCustomerIsLoggedIn').mockReturnValue(true);

    headerWrapper = wrapperFactory();
    headerWrapper.find('[data-testid="link-logout"]').at(0).simulate('click');

    expect(mockLogout).toHaveBeenCalled();
  });
});
