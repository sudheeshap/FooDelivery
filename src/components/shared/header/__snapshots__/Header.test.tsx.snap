// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders the header 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <Header>
        <header
          className="header"
        >
          <nav
            className="navbar"
          >
            <div
              className="navbar__section"
            >
              <div
                className="navbar__item"
              >
                <Link
                  to="/"
                >
                  <LinkAnchor
                    href="/"
                    navigate={[Function]}
                  >
                    <a
                      href="/"
                      onClick={[Function]}
                    >
                      <img
                        alt="Workflow"
                        className="navbar__logo"
                        src="https://tailwindui.com/img/logos/workflow-mark-indigo-500.svg"
                      />
                    </a>
                  </LinkAnchor>
                </Link>
              </div>
            </div>
            <div
              className="navbar__section"
            >
              <div
                className="navbar__item"
              >
                <Link
                  className="navbar__link"
                  data-testid="link-login"
                  to="/login"
                >
                  <LinkAnchor
                    className="navbar__link"
                    data-testid="link-login"
                    href="/login"
                    navigate={[Function]}
                  >
                    <a
                      className="navbar__link"
                      data-testid="link-login"
                      href="/login"
                      onClick={[Function]}
                    >
                      Login
                    </a>
                  </LinkAnchor>
                </Link>
              </div>
              <div
                className="navbar__item"
              >
                <CartIcon
                  count={0}
                  data-testid="link-cart"
                  onClick={[Function]}
                >
                  <div
                    className="cart__icon cart__icon--disabled"
                    onClick={[Function]}
                    onKeyDown={null}
                    role="button"
                    tabIndex="0"
                  >
                    <i
                      className="bi-bag"
                    />
                  </div>
                </CartIcon>
              </div>
            </div>
          </nav>
        </header>
      </Header>
    </Provider>
  </Router>
</BrowserRouter>
`;
